vault/barrier_aes_gcm.go

```diff
@@ -1049,3 +1106,40 @@ func (b *AESGCMBarrier) Keyring() (*Keyring, error) {
 
 	return b.keyring.Clone(), nil
 }
+
+func (b *AESGCMBarrier) encryptTracked(ctx context.Context, keyring *Keyring, path string, term uint32, gcm cipher.AEAD, buf []byte) ([]byte, error) {
+	ct, err := b.encrypt(path, term, gcm, buf)
+	if err != nil {
+		return nil, err
+	}
+	// Increment the local encryption count, and track metrics
+	ops := atomic.AddInt64(&keyring.LocalEncryptions, 1)
+	metrics.IncrCounterWithLabels(barrierEncryptsMetric, 1, termLabel(term))
+
+	// time.Now is expensive, don't check time based rotation on every operation
+	if b.shouldRotate(ops, keyring) {
+		b.l.Lock()
+		defer b.l.Unlock()
+		// Retest the precondition
+		if b.shouldRotate(ops, keyring) {
+			go b.autorotateBarrierKey(ctx)
+		}
+	}
+	return ct, nil
+
+}
+
+func (b *AESGCMBarrier) shouldRotate(ops int64, keyring *Keyring) bool {
+	return ops > b.RotationConfig.KeyRotationMaxOperations ||
+		(b.RotationConfig.KeyRotationInterval > 0 && ops%keyRotationTimeSampleRate == 0 &&
```

Comment:
Should keyRotationTimeSampleRate be configurable via the rotation config?

Reply:
No, that's just an internal configuration.  Trying to pick a value that would avoid calling time.Now() (which requires a system call) on each encryption.  Constant is there because the unit tests use it to do enough writes to trigger the rotation.